File: Chapter02/ch02_homework5.cpp
#include <iostream>
#include <cmath>
#include <iomanip>
#include <array>

// 定义二维向量类型，使用 std::array 比 std::vector 更高效
using Vec2 = std::array<double, 2>;

// 定义梯度函数 ∇f(x, y)
Vec2 gradient_f(double x, double y) {
    return {2.0 * x, 3.0 * y * y};
}

// 计算向量的模长 (L2 Norm)
double norm(const Vec2& v) {
    return std::sqrt(v[0] * v[0] + v[1] * v[1]);
}

int main() {
    // 1. 测试点与梯度计算
    double x0 = 1.0, y0 = 1.0;
    Vec2 grad = gradient_f(x0, y0);
    
    // 梯度方向的单位向量
    double grad_norm = norm(grad);
    Vec2 grad_dir = {grad[0] / grad_norm, grad[1] / grad_norm};
    
    std::cout << "梯度方向单位向量: [" << grad_dir[0] << ", " << grad_dir[1] << "]\n\n";

    // 2. 生成不同方向并计算方向导数
    // 优化：不需要像 NumPy 那样在内存中生成 1000 个向量的数组，
    // 直接在循环中边生成边计算，空间复杂度从 O(N) 降为 O(1)。
    int num_directions = 1000;
    double max_deriv = -1e9; // 初始化为一个极小的值
    double max_theta = 0.0;
    
    // 使用 std::acos(-1.0) 获取 PI，保证在 Windows/Linux 等所有平台下跨平台兼容
    const double PI = std::acos(-1.0); 

    for (int i = 0; i < num_directions; ++i) {
        // 对应 np.linspace(0, 2*np.pi, 1000)
        double theta = 2.0 * PI * i / (num_directions - 1);
        double dx = std::cos(theta);
        double dy = std::sin(theta);
        
        // 计算方向导数 (梯度与方向单位向量的点积)
        double deriv = grad[0] * dx + grad[1] * dy;
        
        // 寻找最大值
        if (deriv > max_deriv) {
            max_deriv = deriv;
            max_theta = theta;
        }
    }

    // 3. 计算角度并输出
    // 计算梯度方向的角度 (弧度转角度)
    // 注意：使用 std::atan2(y, x) 比 std::atan(y/x) 更好，因为它能自动处理四个象限
    double grad_angle_rad = std::atan2(grad_dir[1], grad_dir[0]);
    double grad_angle_deg = grad_angle_rad * 180.0 / PI;
    
    double max_theta_deg = max_theta * 180.0 / PI;

    // 设置输出格式：保留两位小数
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "梯度方向：θ = " << grad_angle_deg << "°\n";
    std::cout << "方向导数最大值对应的角度：θ = " << max_theta_deg << "°\n";

    return 0;
}

File: Chapter02/ch02_homework5.py
# 证明 二元函数函数的梯度方向是函数值上升最快的方向
# 代码验证：以 $f(x, y) = x^2 + y^3$ 为例，在点 $(1, 1)$ 处计算梯度，
# 并比较沿不同方向的方向导数，验证最大值出现在梯度方向。

import numpy as np
import matplotlib.pyplot as plt

# 定义函数及其梯度
def f(x, y):
    return x**2 + y**3

def gradient_f(x, y):
    return np.array([2*x, 3*y**2])

# 测试点
x0, y0 = 1, 1
grad = gradient_f(x0, y0)
grad_dir = grad / np.linalg.norm(grad)  # 梯度方向的单位向量
print(grad_dir)

# 生成不同方向的方向向量（单位向量）
theta = np.linspace(0, 2*np.pi, 1000)
directions = np.column_stack((np.cos(theta), np.sin(theta)))
print(directions)

# 计算方向导数
directional_derivatives = np.sum(grad * directions, axis=1)

# 找到最大值对应的角度
max_idx = np.argmax(directional_derivatives)
max_theta = theta[max_idx]

# 输出结果（用反三角函数arctan，然后弧度转角度）
print(f"梯度方向：θ = {np.degrees(np.arctan(grad_dir[1]/grad_dir[0])):.2f}°")
print(f"方向导数最大值对应的角度：θ = {np.degrees(max_theta):.2f}°")


# 绘制方向导数随角度的变化
plt.figure(figsize=(8, 5))
plt.plot(np.degrees(theta), directional_derivatives, label="function")
plt.axvline(np.degrees(max_theta), color='r', linestyle='--', label="max")
plt.xlabel("θ°")
plt.ylabel("y")
plt.legend()
plt.grid(True)
plt.show()
plt.savefig('./第2章 机器学习的数学基础/ch02_homework5.png')